# Prueba Intertrimestral

**Nombre:**

**Apellidos:**

**Tiempo de la prueba: 2 Horas**

**Asignatura:** Desarrollo de Aplicaciones para la Visualización de Datos

**Fecha:** 18 de octubre de 2023

**Instrucciones:**

- Escribe código limpio y autoexplicativo.

- Se eliminará 0.5 puntos por usar Seaborn o Matplotlib.

- Se pueden utilizar los materiales de clase.

- Se puede utilizar internet para búsqueda de dudas y documentación.

- No se puede utilizar ningún tipo de LLM.

- No se puede utilizar mensajería instantánea.

- Sube tus resultados a tu repositorio de Github.

- Imprime una versión en PDF en A3 y Portrait del notebook.

- Envialo tus resultados a dmartincorral@icai.comillas.edu adjuntando el PDF y la url del notebook subido al repositorio de Github.


## Inicialización de librerías

Carga aquí todas las librerías que vayas a utilizar.

In [1]:
import datetime
import numpy as np
import pandas as pd
from datetime import date, timedelta, datetime
import sklearn.datasets
import matplotlib.pyplot as plt
import seaborn as sns
from math import sqrt


from sklearn.model_selection import train_test_split
from sklearn.svm import LinearSVC
from sklearn.metrics import (adjusted_mutual_info_score, homogeneity_score,
                            completeness_score,classification_report, confusion_matrix,
                            mean_squared_error, mean_absolute_error,
                            mean_absolute_percentage_error,
                            silhouette_score, v_measure_score, adjusted_rand_score)
from sklearn.linear_model import ElasticNet, SGDClassifier, LogisticRegression
from sklearn.cluster import KMeans, DBSCAN, AffinityPropagation

import plotly.graph_objects as go
import plotly.express as px
import statsmodels.api as sm

from imblearn.over_sampling import RandomOverSampler
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


## Ejercicio 1 (2 puntos):

a) Crea una función que calcule y devuelva el factorial de un número entero. **(0.6 puntos)**

b) Crea una función que verifique si un número es primo o no. **(0.6 puntos)**

c) Muestra en un dataframe los 50 primeros números positivos, si es primo y su factorial utilizando las funciones anteriores. **(0.6 puntos)**

d) ¿Cómo se podría programar en una clase las tres operaciones anteriores? **(0.2 puntos)**

In [13]:
# a) Crea una función que calcule y devuelva el factorial de un número entero. **(0.6 puntos)**

def Factorial(n: int):
	if n < 0:
		return 0
	elif n == 0 or n == 1:
		return 1
	else:
		fact = 1
		while(n > 1):
			fact *= n
			n -= 1
		return fact

# Driver Code
num = 5
print("El Factorial de",num,"es",Factorial(num))


Factorial of 5 is 120


In [21]:
#b) Crea una función que verifique si un número es primo o no. **(0.6 puntos)**

def esPrimo(n):
    if (n==1):
        return False
    elif (n==2):
        return True;
    else:
        for x in range(2,n):
            if(n % x==0):
                return False
        return True

print("Es primo el ",num,"?",esPrimo(9))


Es primo el  5 ? False


In [ ]:
#c) Muestra en un dataframe los 50 primeros números positivos, si es primo y su factorial utilizando las funciones anteriores. **(0.6 puntos)**
df = pd.DataFrame()
df.insert(0, 'Numeros', range(1,51))
df['Factorial'] = df['Numeros'].apply(Factorial)
df['Primo'] = df['Numeros'].apply(esPrimo)
df

In [23]:
#d) ¿Cómo se podría programar en una clase las tres operaciones anteriores? **(0.2 puntos)**
class Entero():
    def __init__(self,n):
      if isinstance(n, int):
        self.n=n
      else:
        raise ValueError("No es entero")

    def Factorial(self):
      if self.n < 0:
        return 0
      elif self.n == 0 or self.n == 1:
        return 1
      else:
        fact = 1
        while(self.n > 1):
          fact *= self.n
          self.n -= 1
        return fact

    def esPrimo(self):
        if (self.n==1):
            return False
        elif (self.n==2):
            return True;
        else:
            for x in range(2,self.n):
                if(self.n % x==0):
                    return False
            return True

prueba = Entero(3)
print(prueba.Factorial())
print(prueba.esPrimo())

6
False


## Ejercicio 2 (4 puntos):

a) Extrae de sklearn el conjunto de datos __California Housing dataset__
 y transfórmalo a dataframe de pandas **(0.25 puntos)**

b) Construye una función que muestra la estructura del dataset, el número de NAs, tipos de variables y estadísticas básicas de cada una de las variables. **(0.5 puntos)**

c) Construye una __Regresión lineal__ y un __Random forest__ que predigan el __Median house value__ según los datos disponibles. **(0.75 puntos)**

d) Visualiza cuales son las variables (coeficientes) más importantes en cada uno de los modelos. **(1.25 puntos)**

e) Decide a través de las métricas que consideres oportunas, cuál de los dos modelos es mejor, por qué y explica el proceso que has realizado para responder en los puntos anteriores. **(1.25 puntos)**

In [25]:
#a) Extrae de sklearn el conjunto de datos __California Housing dataset__  y transfórmalo a dataframe de pandas **(0.25 puntos)**

from sklearn.datasets import fetch_california_housing

california = fetch_california_housing()

import pandas as pd

california_df = pd.DataFrame(california.data,
                             columns=california.feature_names)
california_df['MedHouseValue'] = pd.Series(california.target)
california_df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  MedHouseValue  
0    -122.23          4.526  
1    -122.22          3.585  
2    -122.24          3.521  
3    -122.25          3.413  
4    -122.25          3.422

In [94]:
#b) Construye una función que muestra la estructura del dataset, el número de NAs, tipos de variables y estadísticas básicas de cada una de las variables. **(0.5 puntos)**
def infoDataset(df: pd):
  print("Informacion\n\n")
  print(df.info())
  print("#############################")
  print("Valores nulos\n\n",df.isnull().sum())
  print("#############################")
  print("Tipos de variables\n\n",df.dtypes)
  print("#############################")
  print("Estadisticas basicas de las variables\n\n",df.describe())
  # Check for missing values
  missing_values = df.isnull().sum()
  print("#############################")
  print("MISSING VALUES")

  print(missing_values[missing_values > 0])

infoDataset(california_df)


Informacion


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   MedInc         20640 non-null  float64
 1   HouseAge       20640 non-null  float64
 2   AveRooms       20640 non-null  float64
 3   AveBedrms      20640 non-null  float64
 4   Population     20640 non-null  float64
 5   AveOccup       20640 non-null  float64
 6   Latitude       20640 non-null  float64
 7   Longitude      20640 non-null  float64
 8   MedHouseValue  20640 non-null  float64
dtypes: float64(9)
memory usage: 1.4 MB
None
#############################
Valores nulos

 MedInc           0
HouseAge         0
AveRooms         0
AveBedrms        0
Population       0
AveOccup         0
Latitude         0
Longitude        0
MedHouseValue    0
dtype: int64
#############################
Tipos de variables

 MedInc           float64
HouseAge         float64
AveRooms         f

In [44]:
#c) Construye una Regresión lineal y un Random forest que predigan el Median house value según los datos disponibles. (0.75 puntos)



from sklearn.model_selection import train_test_split

X = california_df.drop('MedHouseValue', axis=1)
y = california_df['MedHouseValue']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [60]:
#### REGRESION LINEAL
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Inicializar y entrenar el modelo
model = LinearRegression()
model.fit(X_train, y_train)

# Predicciones
y_pred = model.predict(X_test)

# Evaluación
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

# Use the forest's predict method on the test data
predictions = model.predict(X_test)
# Calculate the absolute errors
errors = abs(predictions - y_test)
# Print out the mean absolute error (mae)
print('Mean Absolute Error:', round(np.mean(errors), 2), 'degrees.')
# Calculate mean absolute percentage error (MAPE)
mape = 100 * (errors / y_test)
# Calculate and display accuracy
accuracy = 100 - np.mean(mape)
print('Accuracy:', round(accuracy, 2), '%.')


print(f"MSE (Mean Squared Error): {mse}")
print(f"R^2 Score: {r2}")

# plot de valores reales vs predicciones
import plotly.express as px

fig = px.scatter(x=y_test, y=y_pred,opacity=0.65,
    trendline='ols', trendline_color_override='darkblue',
                 labels={
                     "x": "Valores reales",
                     "y": "Predicciones",
                 },
                title="Valores reales vs Predicciones")

fig.show()


Mean Absolute Error: 0.53 degrees.
Accuracy: 68.05 %.
MSE (Mean Squared Error): 0.5558915986952444
R^2 Score: 0.5757877060324508


In [62]:
#Random forest

# Import the model we are using
from sklearn.ensemble import RandomForestRegressor
# Instantiate model with 100 decision trees
rf = RandomForestRegressor(n_estimators=100)
# Train the model on training data
rf.fit(X_train, y_train);

# Use the forest's predict method on the test data
predictions = rf.predict(X_test)
# Calculate the absolute errors
errors = abs(predictions - y_test)
# Print out the mean absolute error (mae)
print('Mean Absolute Error:', round(np.mean(errors), 2), 'degrees.')

# Calculate mean absolute percentage error (MAPE)
mape = 100 * (errors / y_test)
# Calculate and display accuracy
accuracy = 100 - np.mean(mape)
print('Accuracy:', round(accuracy, 2), '%.')



Mean Absolute Error: 0.33 degrees.
Accuracy: 81.09 %.


tras varias pruebas he decido poner estos valores en el modelo para no tener overfitting y tambien por problemas de tiempo de procesado
n_estimators=100
Vemos que el accurary es muy superior al linear regression.

Linear regression - Accuracy: 68.05 %.
Random forest - Accuracy: 81.13 %.


In [85]:
# d) Visualiza cuales son las variables (coeficientes) más importantes en cada uno de los modelos. (1.25 puntos)
# LINEAR REGRESSION
# Get numerical feature importances
importances = list(model.coef_)
# List of tuples with variable and importance
feature_importances = [(feature, round(importance, 2)) for feature, importance in zip(X.columns, importances)]
# Sort the feature importances by most important first
feature_importances = sorted(feature_importances, key = lambda x: x[1], reverse = True)
# Print out the feature and importances
[print('Variable: {:20} Importance: {}'.format(*pair)) for pair in feature_importances];

fig = px.bar(x=X.columns, y=importances ,
             title="Importancia de las variables en Linear Regression")

fig.show()


Variable: AveBedrms            Importance: 0.78
Variable: MedInc               Importance: 0.45
Variable: HouseAge             Importance: 0.01
Variable: Population           Importance: -0.0
Variable: AveOccup             Importance: -0.0
Variable: AveRooms             Importance: -0.12
Variable: Latitude             Importance: -0.42
Variable: Longitude            Importance: -0.43


Vemos que la variable mas importante es la media de habitaciones. La cual tiene una relacion positiva con el valor de la casa.

In [79]:
# RANDOM FOREST
# Get numerical feature importances
importances = list(rf.feature_importances_)
# List of tuples with variable and importance
feature_importances = [(feature, round(importance, 2)) for feature, importance in zip(X.columns, importances)]
# Sort the feature importances by most important first
feature_importances = sorted(feature_importances, key = lambda x: x[1], reverse = True)
# Print out the feature and importances
[print('Variable: {:20} Importance: {}'.format(*pair)) for pair in feature_importances];

fig = px.bar(x=X.columns, y=importances ,
             title="Importancia de las variables en Random Forest")

fig.show()

Variable: MedInc               Importance: 0.53
Variable: AveOccup             Importance: 0.14
Variable: Latitude             Importance: 0.09
Variable: Longitude            Importance: 0.09
Variable: HouseAge             Importance: 0.05
Variable: AveRooms             Importance: 0.04
Variable: AveBedrms            Importance: 0.03
Variable: Population           Importance: 0.03


Vemos que la variable mas importante es el media salario medio. Lo cual puede ser una pregunta clave en el arbol de decisiones, para ayer la mejor prediccion.

In [ ]:
#e) Decide a través de las métricas que consideres oportunas, cuál de los dos modelos es mejor,
# por qué y explica el proceso que has realizado para responder en los puntos anteriores. (1.25 puntos)


El modelo que funciona mejor en mi opinion es Random Forest, ya que es el que mayor accuracy tiene.
Lineal regression tiene un Accuracy de 68.05 %, sin embargo Random forest tiene un Accuracy de 81.13 %. Ademas se ha limitado el modelo del random forest para evitar posibles overfittins. Ademas viendo las variables mas importante vemos que el salario medio es clave en el proceso de regression de este ultimo modelo, lo cual tiene mucho sentido, ya que suele haber una relacion entre el salario del dueño y el precio de la casa.

## Ejercicio 3 (4 puntos):


*Consideremos* el dataset que contiene __The Most Streamed Spotify Songs 2023__ que se encuentra en el respositorio.

Información de las variables:

- track_name: Name of the song
- artist(s)_name: Name of the artist(s) of the song
- vartist_count: Number of artists contributing to the song
- released_year: Year when the song was released
- released_month: Month when the song was released
- release_day: Day of the month when the song was released
- in_spotify_playlists: Number of Spotify playlists the song is included in
- in_spotify_charts: Presence and rank of the song on Spotify charts
- streams: Total number of streams on Spotify
- in_apple_playlists: Number of Apple Music playlists the song is included in
- in_apple_charts: Presence and rank of the song on Apple Music charts
- in_deezer_playlists: Number of Deezer playlists the song is included in
- in_deezer_charts: Presence and rank of the song on Deezer charts
- in_shazam_charts: Presence and rank of the song on Shazam charts
- bpm: Beats per minute, a measure of song tempo
- key: Key of the song
- mode: Mode of the song (major or minor)
- danceability_%: Percentage indicating how suitable the song is for dancing
- valence_%: Positivity of the song's musical content
- energy_%: Perceived energy level of the song
- acousticness_%: Amount of acoustic sound in the song
- instrumentalness_%: Amount of instrumental content in the song
- liveness_%: Presence of live performance elements
- speechiness_%: Amount of spoken words in the song

Para las respuestas b, c, d, e, f y g es imperativo acompañarlas respuestas con una visualización.

a) Lee el fichero en formato dataframe, aplica la función del ejercicio 2.b, elimina NAs y convierte a integer si fuera necesario. **(0.25 puntos)**

b) ¿Cuántos artistas únicos hay? **(0.25 puntos)**

c) ¿Cuál es la distribución de reproducciones? **(0.5 puntos)**

d) ¿Existe una diferencia signitificativa en las reproducciones entre las canciones de un solo artista y las de más de uno? **(0.5 puntos)**

e) ¿Cuáles son las propiedades de una canción que mejor correlan con el número de reproducciones de una canción? **(0.5 puntos)**

f) ¿Cuáles son las variables que mejor predicen las canciones que están por encima el percentil 50? **(1 puntos)**

*Nota: Crea una variable binaria (Hit/No Hit) en base a 3.c, crea una regresión logística y visualiza sus coeficientes.*

g) Agrupa los 4 gráficos realizados en uno solo y haz una recomendación a un sello discográfico para producir un nuevo hit. **(1 puntos)**


In [95]:
#a) Lee el fichero en formato dataframe, aplica la función del ejercicio 2.b, elimina NAs y convierte a integer si fuera necesario. **(0.25 puntos)**
spotify = pd.read_csv('./drive/MyDrive/Desarrollo de Aplicaciones/Inter/spotify-2023.csv', encoding='ISO-8859-1')
infoDataset(spotify)


Informacion


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 953 entries, 0 to 952
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   track_name            953 non-null    object
 1   artist(s)_name        953 non-null    object
 2   artist_count          953 non-null    int64 
 3   released_year         953 non-null    int64 
 4   released_month        953 non-null    int64 
 5   released_day          953 non-null    int64 
 6   in_spotify_playlists  953 non-null    int64 
 7   in_spotify_charts     953 non-null    int64 
 8   streams               953 non-null    object
 9   in_apple_playlists    953 non-null    int64 
 10  in_apple_charts       953 non-null    int64 
 11  in_deezer_playlists   953 non-null    object
 12  in_deezer_charts      953 non-null    int64 
 13  in_shazam_charts      903 non-null    object
 14  bpm                   953 non-null    int64 
 15  key                   858 

Vemos que hay nan en key y en in_shazam_chats, por lo que lo elimines, el cambiarlo a integer lo vere sobretodo cuando tenga que visualizar si hace falta


In [149]:
spotify.dropna(inplace=True)
spotify["streams"] = pd.to_numeric(spotify["streams"], errors='coerce')


In [114]:
#b) ¿Cuántos artistas únicos hay? (0.25 puntos)
print(spotify["artist(s)_name"].unique())
# no se para las colaboraciones

['Latto, Jung Kook' 'Myke Towers' 'Olivia Rodrigo' 'Taylor Swift'
 'Bad Bunny' 'Dave, Central Cee' 'Eslabon Armado, Peso Pluma' 'Quevedo'
 'Gunna' 'Peso Pluma, Yng Lvcas' 'Bad Bunny, Grupo Frontera' 'NewJeans'
 'David Kushner' 'SZA' 'Fifty Fifty' 'Feid, Young Miko' 'Jimin'
 'Gabito Ballesteros, Junior H, Peso Pluma' 'Bizarrap, Peso Pluma'
 'The Weeknd, Madonna, Playboi Carti' 'Fuerza Regida'
 'R��ma, Selena G' 'Tainy, Bad Bunny' 'Morgan Wallen' 'Dua Lipa'
 'Troye Sivan' 'Peso Pluma, Grupo Frontera'
 'The Weeknd, 21 Savage, Metro Boomin' 'Karol G, Shakira'
 'Yahritza Y Su Esencia, Grupo Frontera' 'Junior H, Peso Pluma'
 'Post Malone, Swae Lee' 'Bebe Rexha, David Guetta'
 'Tyler, The Creator, Kali Uchis' 'Miley Cyrus' 'The Weeknd'
 'The Weeknd, Daft Punk' 'Ariana Grande, The Weeknd'
 'Feid, Myke Towers, Sky Rompiendo' 'David Guetta, Anne-Marie, Coi Leray'
 'Karol G' 'Peggy Gou' 'Kali Uchis' 'Manuel Turizo'
 'dennis, MC Kevin o Chris' 'PinkPantheress, Ice Spice'
 'Charlie Puth, BTS, J

In [131]:
#c) ¿Cuál es la distribución de reproducciones? (0.5 puntos)
# Histograma interactivo con `plotly.express`
fig1 = px.histogram(spotify, x='streams', marginal='box')
fig1.show()

In [118]:
#d) ¿Existe una diferencia signitificativa en las reproducciones entre las canciones de un solo artista y las de más de uno? (0.5 puntos)
fig2 = px.histogram(spotify, x="streams", color="artist_count")
fig2.show()


vemos que las de dos artistas son las que peor funcionan, el resto son mas parecidas.

In [129]:
#e) ¿Cuáles son las propiedades de una canción que mejor correlan con el número de reproducciones de una canción? (0.5 puntos)
spotify_corr = spotify.corr()

fig = go.Figure()
fig.add_trace(
    go.Heatmap(
        x = spotify_corr.columns,
        y = spotify_corr.index,
        z = np.array(spotify_corr),
        text=spotify_corr.values,
        texttemplate='%{text:.2f}'
    )
)
fig.show()

<ipython-input-129-d577a0718b7f>:2: FutureWarning:

The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.



nos fijamos en la variables streams y vemos que tiene mucha correlacion con in_spotify_playlists y con in_apple_playlists, lo que nos indica la gran importancia de estar en playlists para conseguir mas visibilidad, ademas de que si esta en una playlist es mas probable que una persona la escuche varias veces.

In [181]:
#f) ¿Cuáles son las variables que mejor predicen las canciones que están por encima el percentil 50? (1 puntos)
spotify_filtrado = spotify[spotify.streams > np.percentile(spotify.streams, 50)]
fig7 = px.histogram(spotify_filtrado, x='streams', marginal='box')
fig7.show()


In [182]:
from sklearn.model_selection import train_test_split

spotify_filtrado.drop(['artist(s)_name','track_name'],inplace=True,axis=1)

X = spotify_filtrado.drop('streams', axis=1)
y = spotify_filtrado['streams']

X = X.apply(pd.to_numeric, errors='coerce')
y = y.apply(pd.to_numeric, errors='coerce')

X.fillna(0, inplace=True)
y.fillna(0, inplace=True)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


<ipython-input-182-537ac4979a93>:5: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [183]:
#### REGRESION LINEAL
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Inicializar y entrenar el modelo
model = LinearRegression()
model.fit(X_train, y_train)

# Predicciones
y_pred = model.predict(X_test)

# Evaluación
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

# Use the forest's predict method on the test data
predictions = model.predict(X_test)
# Calculate the absolute errors
errors = abs(predictions - y_test)
# Print out the mean absolute error (mae)
print('Mean Absolute Error:', round(np.mean(errors), 2), 'degrees.')
# Calculate mean absolute percentage error (MAPE)
mape = 100 * (errors / y_test)
# Calculate and display accuracy
accuracy = 100 - np.mean(mape)
print('Accuracy:', round(accuracy, 2), '%.')


print(f"MSE (Mean Squared Error): {mse}")
print(f"R^2 Score: {r2}")

# plot de valores reales vs predicciones
import plotly.express as px

fig = px.scatter(x=y_test, y=y_pred,opacity=0.65,
    trendline='ols', trendline_color_override='darkblue',
                 labels={
                     "x": "Valores reales",
                     "y": "Predicciones",
                 },
                title="Valores reales vs Predicciones")

fig.show()

# d) Visualiza cuales son las variables (coeficientes) más importantes en cada uno de los modelos. (1.25 puntos)
# LINEAR REGRESSION
# Get numerical feature importances
importances = list(model.coef_)
# List of tuples with variable and importance
feature_importances = [(feature, round(importance, 2)) for feature, importance in zip(X.columns, importances)]
# Sort the feature importances by most important first
feature_importances = sorted(feature_importances, key = lambda x: x[1], reverse = True)
# Print out the feature and importances
[print('Variable: {:20} Importance: {}'.format(*pair)) for pair in feature_importances];

fig = px.bar(x=X.columns, y=importances ,
             title="Importancia de las variables en Linear Regression")

fig.show()



Mean Absolute Error: 239619295.29 degrees.
Accuracy: 58.66 %.
MSE (Mean Squared Error): 1.0834041019851728e+17
R^2 Score: 0.5984395859803231


Variable: in_spotify_charts    Importance: 6308238.05
Variable: released_year        Importance: 5664318.93
Variable: released_day         Importance: 3580464.96
Variable: in_apple_playlists   Importance: 2390017.07
Variable: acousticness_%       Importance: 2226210.12
Variable: liveness_%           Importance: 336706.93
Variable: in_deezer_playlists  Importance: 137745.8
Variable: in_spotify_playlists Importance: 30852.98
Variable: key                  Importance: 0.0
Variable: mode                 Importance: 0.0
Variable: speechiness_%        Importance: -203386.78
Variable: energy_%             Importance: -621949.12
Variable: in_apple_charts      Importance: -674660.88
Variable: in_shazam_charts     Importance: -675204.16
Variable: bpm                  Importance: -984057.39
Variable: instrumentalness_%   Importance: -1143143.28
Variable: valence_%            Importance: -1299317.28
Variable: danceability_%       Importance: -2158418.77
Variable: in_deezer_charts     Importance: -

Vemos que la variable mas importante es artist count, esto puede significar que cuando las reproduccion ya son altas(encima del percentil 50), la cuantos mas artistas menos en una cancion hay menos reproducciones hay, esto lo podemos ver porque el coef artist count es negativo.

In [184]:
#Random forest

# Import the model we are using
from sklearn.ensemble import RandomForestRegressor
# Instantiate model with 100 decision trees
rf = RandomForestRegressor(n_estimators=100)
# Train the model on training data
rf.fit(X_train, y_train);

# Use the forest's predict method on the test data
predictions = rf.predict(X_test)
# Calculate the absolute errors
errors = abs(predictions - y_test)
# Print out the mean absolute error (mae)
print('Mean Absolute Error:', round(np.mean(errors), 2), 'degrees.')

# Calculate mean absolute percentage error (MAPE)
mape = 100 * (errors / y_test)
# Calculate and display accuracy
accuracy = 100 - np.mean(mape)
print('Accuracy:', round(accuracy, 2), '%.')



Mean Absolute Error: 201480496.4 degrees.
Accuracy: 69.61 %.


In [185]:
# RANDOM FOREST
# Get numerical feature importances
importances = list(rf.feature_importances_)
# List of tuples with variable and importance
feature_importances = [(feature, round(importance, 2)) for feature, importance in zip(X.columns, importances)]
# Sort the feature importances by most important first
feature_importances = sorted(feature_importances, key = lambda x: x[1], reverse = True)
# Print out the feature and importances
[print('Variable: {:20} Importance: {}'.format(*pair)) for pair in feature_importances];

fig = px.bar(x=X.columns, y=importances ,
             title="Importancia de las variables en Random Forest")

fig.show()

Variable: in_spotify_playlists Importance: 0.63
Variable: released_year        Importance: 0.06
Variable: in_apple_playlists   Importance: 0.06
Variable: in_deezer_playlists  Importance: 0.04
Variable: energy_%             Importance: 0.03
Variable: in_spotify_charts    Importance: 0.02
Variable: in_apple_charts      Importance: 0.02
Variable: bpm                  Importance: 0.02
Variable: danceability_%       Importance: 0.02
Variable: valence_%            Importance: 0.02
Variable: acousticness_%       Importance: 0.02
Variable: released_month       Importance: 0.01
Variable: released_day         Importance: 0.01
Variable: in_deezer_charts     Importance: 0.01
Variable: in_shazam_charts     Importance: 0.01
Variable: liveness_%           Importance: 0.01
Variable: speechiness_%        Importance: 0.01
Variable: artist_count         Importance: 0.0
Variable: key                  Importance: 0.0
Variable: mode                 Importance: 0.0
Variable: instrumentalness_%   Importance: 

Sin embargo vemos que para el random forest el estar en listas de spotify sigue siendo una variable cable, como vimos en la matriz de correlacion.